### DISCUSSION TITLE
About lifetimes

### DISCUSSION THREAD

**Discussion by f-saez** [modular/modular#3189] [Q&A] [answered]
I'm playing with FFI and I need to know how to manage correctly some lifetimes. 
Let's take an example :
```
alias external_func = fn(UnsafePointer[JpegErrorMgr]) -> UInt

@value
struct Stuff:
  var pointer : UnsafePointer[UInt8]
  var ... # many other fields

fn main():
  var thing = ffi.DLHandle("liblambda.so", ffi.RTLD.NOW)
  if thing.__bool__():
    var bytes = List[UInt8]()
    with open("file", "rb") as f:
      bytes = f.read_bytes()  

var my_struct = Stuff()
my_struct.pointer = bytes.unsafe_ptr()
var result = thing.get_function[external_func]("external_func")(my_struct.unsafe_ptr()) 
# others call to external_func but no other use of my_struct
```    
    
If I understand correctly, for Mojo, the life of bytes stops as soon as my_struct is not used and in this case, it's just after the first call to "external_func". Trouble is external_func will need bytes to be alive for a longer time but Mojo can't know that and external_func will starts to act funny because the values of bytes may have changed

So, my question is : how to manage this kind of lifetime ? an easy answer is to do some useless stuff with my_struct at the end of the block  (it's what I do now), but a cleaner way to do that must exists.

**Comment by Brian-M-J** [comment]
One solution is to put a `_ = my_struct` line after all the uses of `external_func`. If you don't want to do that, you can create a context manager which automatically keeps the `my_struct` alive for the entire block.

**Comment by f-saez** [comment, marked as answer]
great ! Thanks for the answer.
